[Video] Func spec update & migrate (#1060)
carbon-design-system/carbon-for-ibm-dotcom-website: wiki/UI-component/Video.md

<details open="true">
  <summary><strong>Dependencies</strong> (<!-- dependencyCount start --><!-- dependencyCount end -->)</summary><br />

- [[Video service]]
- [[Lightbox media viewer]]
- [[Image]]

> - Kaltura player

Reviewer: Can we reference `r2`?

<details open="true">
  <summary><strong>Resources</strong></summary><br />

- r1: [Box folder](https://ibm.box.com/s/9q7rdj73m32hufcw1xwl1td6nr88rghm)
- r2: [Kaltura player](http://player.kaltura.com/docs/api)

<br />
</details>

<hr width="40%" />

## Functions

| Variants | Interaction | Functions |
| -------- | ----------- | --------- |
| Default | Clicking on `Poster image` | Poster image is replaced with Kaltura video player and video will begin playing. |  

> | with lightbox media viewer | Clicking on `Poster image` | Open LightBoxMediaViewer and video will begin playing. |

Reviewer: I feel like this basically repeats the information you have below.

## Logics

> ### Default variant

Reviewer: Why do we have this level of headings? I don't know if it's necessary, and it doesn't really follow how we're doing other components.

Video allows custom override of all content returned by the [[Video service]] except video duration.

| ID  | Field |
| --- | ----- |
| 1   | Caption |
| 2   | Poster image |
| 3   | Poster image alt tag |

> ##### 1. Caption

Reviewer: I try to follow the data model if possible. I would call this out as copy personally.

| Condition | Description |
| --------- | ----------- |
| `undefined` | Video name prefilled by [[Video service]]. Duration is appended after name in parenthesis. |

> | Text | Adding a custom caption overrides the API call. Duration is appended after custom caption in parenthesis. |

Reviewer: I would personally add more of a template of how it's intended instead of being so verbose. Or not describe how it should appear at all since that's where design comes in more. Honestly, your first sentence of each row might be good enough here.

```
video.title (video.duration)
```

##### 2. Poster image

Poster image should always obeys aspect ratios: 16x9, 9x16, 2x1, 1x2, 4x3, 3x4, 1x1.

| Condition | Description |
| --------- | ----------- |
| `undefined` | Poster image prefilled by [[Video service]] |

> | Image | Adding a custom poster image overrides the API call. |

Reviewer: - We should call out we're using the Image component here.
- Title should be `Image`.
- I would put ratios in a table as an option or condition.
- table is missing ###### Logic heading.

##### 3. Poster image alt tag

| Condition | Description |
| --------- | ----------- |
| `undefined` | Video description prefilled by [[Video service]]. |

> | Text | Adding a custom alt tag description overrides the API call. |

Reviewer: This is in documented in the Image functional specs.

> ### with Lightbox media viewer variant

Reviewer: I would avoid using the word "variants" and "with" because there are so many variants here you're not calling out as variants.

> ### with Lightbox media viewer variant

Reviewer: I would personally put all of this as the level 3 header to follow the other functional specs consistently.

| ID  | Field |
| --- | ----- |
| 4   | Heading |

> | 5   | Body copy |

Reviewer: Duplicating content below, so I would argue unecessary.

##### 4. Heading

| Condition | Description |
| --------- | ----------- |

> | `undefined` | Video name prefilled by [[Video service]]. |

Reviewer: Might would state somewhere that these are the default states. Developers like to see that more clearly sometimes.

| Text | Adding a custom heading overrides the API call. |

> ##### 5. Body copy

Reviewer: Just `Copy` to align with everything else.

| Condition | Description |
| --------- | ----------- |

> | `undefined` | Video description prefilled by [[Video service]]. |

Reviewer: We should call out what field by the video service if we know it. `video.description`. Let me know if you need help tracking this information down.

> | `undefined` | Video description prefilled by [[Video service]]. |

Reviewer: Also, this isn't the video description. It's the lightbox media viewer copy, and it's prefilled by the video description.

> 

Reviewer: I think we need to add [Image](https://github.com/carbon-design-system/carbon-for-ibm-dotcom-website/wiki/Image) as a dependency here.

<hr width="40%" />

> ## 1. Data & content

Reviewer: This I believe is starting to get into the territory of the video service in my opinion. Which we don't have a functional specs for because it kept getting put on the back burner.

This appears to me to be a list of data provided by the Kaltura API, instead of the data and content model of this component. Am I understanding your intent here correctly?

Author: list of data provided by the Kaltura API and content model of this component seems related. Maybe I can restructure to focus / lead with on content of this component.

> The video component intakes a Kaltura `videoId` and fetches the following data to be rendered on page as content. All of these content can be customized except the video duration.

Reviewer: Try to reference back to the resources when you call them out.

```suggestion
The video component intakes a Kaltura player <sup>[[r2](#resources)]</sup> `videoId` and fetches the following data to be rendered on page as content. All of these content can be customized except the video duration.
```

> See more details at [Kaltura Player API](http://player.kaltura.com/docs/api).

Reviewer: 1. Try to reference back to the resources when you call them out.
2. Use the callouts from the functional specs template for moments like these.

```suggestion
> ![image](https://user-images.githubusercontent.com/3793636/117873919-f6faba80-b265-11eb-81a5-039bdcd822e8.png)  Check out the Kaltura player <sup>[[r2](#resources)]</sup> for more information
```

| Data name | Type | Allows override |
| --------- | ---- | --------------- |
| Name | Text | Yes |
| Description | Text | Yes |
| Duration | Number | No |
| Poster image | Image | Yes |

## 2. Rendering & behavior

> Video is used in the following components. Depending on the visual design and content requirement of the final component rendered, the above data is used differently.

Reviewer: Cross link when possible. Even if it's back to this same page.

```suggestion
[[Video]] is used in the following components. Depending on the visual design and content requirement of the final component rendered, the above data is used differently.
```

### Video

> This is the default rendering of video. It is used for embedding a video directly on page or inline.

Reviewer: 1. Cross link when possible. Even if it's back to the same page.
2. It's less of embedding a video, and more of embedding the Kaltura player
3. Nit picky thing, but inline is the same as directly in the page in my opinion.

```suggestion
This is the default rendering of [[Video]]. It is used for embedding a Kaltura player <sup>[[r2](#resources)]</sup> directly in the page.
```

![video](https://user-images.githubusercontent.com/15144993/123469683-5cf69380-d5c1-11eb-844f-39e26678c961.png)

| ID  | Data name     | Field |
| --- | ------------- | ----- |
| 1   | Name   | Caption |
| 2   | Duration | appended after caption in parenthesis |
| 3   | Description | Poster image alt tag |
| 4   | Poster image | Image |

Interactions

| ID  | Element name | Event | Behavior | Notes |
| --- | ------------ | ----- | -------- | ----- |

> | 1   | Poster image | On Click or On Enter key of link | Poster image will be replaced by Kaltura video player that will begin playing <br /> Alternative option to open in LightBoxMediaViewer. | Poster image obeys aspect ratios: 16x9, 9x16, 2x1, 1x2, 4x3, 3x4, 1x1 |

Reviewer: Scoping it all back. This is where your functional specs for this component begin. Everything around this belongs in something else.

### CTA - Text

![CTA - Text](https://user-images.githubusercontent.com/15144993/123468111-523aff00-d5bf-11eb-8ef1-2f05c0de29c6.png)

| ID  | Data name     | Field |
| --- | ------------- | ----- |
| 1   | Name | CTA text |
| 2   | Duration | appended after CTA text in parenthesis |

Interactions

| ID  | Element name | Event | Behavior |
| --- | ------------ | ----- | -------- |
| 1   | CTA - Text | On Click or On Enter key of link | Open video in LightBoxMediaViewer. |

### CTA - Button

![CTA - Button](https://user-images.githubusercontent.com/15144993/123468110-523aff00-d5bf-11eb-9b1d-0c65f082d482.png)

> | ID  | Data name     | Field |

Reviewer: Missing a title as to what this table is here for.

```suggestion
###### Data
| ID  | Data name     | Field |
```

| --- | ------------- | ----- |
| 1   | Name | Button label |
| 2   | Duration | appended after Button label in parenthesis |

> Interactions

Reviewer: 1. Follow the functional spec template hiearchy when possible.
2. I'd also suggest Logic as the title to align with the rest of the specs.

```suggestion
###### Logic
```

| ID  | Element name | Event | Behavior |
| --- | ------------ | ----- | -------- |
| 1   | CTA - Button | On Click or On Enter key of button | Open video in LightBoxMediaViewer. |

### CTA - Card

![CTA - Card](https://user-images.githubusercontent.com/15144993/123468108-51a26880-d5bf-11eb-93a0-c23ef1c8216b.png)

> | ID  | Data name     | Field |

Reviewer: Missing a title

```suggestion
###### Logic
| ID  | Data name     | Field |
```

| --- | ------------- | ----- |
| 1   | Name | Card text |
| 2   | Description | Card body copy |
| 3   | Duration | CTA text | No |
| 4   | Poster image | Image |

> Interactions

Reviewer: 1. Follow template headings
2. I tend to call sections like this `Logic`.

```suggestion
###### Logic
```

| ID  | Element name | Event | Behavior |
| --- | ------------ | ----- | -------- |

> | 1   | CTA - Card   | On Click or On Enter key of card | Open video in LightBoxMediaViewer. |

Reviewer: I believe this type of information should belong over in [CTA](https://github.com/carbon-design-system/carbon-for-ibm-dotcom-website/wiki/CTA).

### Lightbox media viewer

![Lightbox](https://user-images.githubusercontent.com/15144993/123468106-5109d200-d5bf-11eb-8556-828dc20b5ccd.png)

| ID  | Data name     | Field |
| --- | ------------- | ----- |
| 1   | Name | Lightbox headline |
| 2   | Description | Lightbox body copy |
| 3   | Poster image | Image |

Interactions

| ID  | Element name | Event | Behavior |
| --- | ------------ | ----- | -------- |
| 1   | LightboxMediaViewer | On open | Begins playing video |
| 2   | Kaltura embedded video | See [standard Kaltura Player functionality](http://player.kaltura.com/docs/main) | Contains all available functionality of the standard Kaltura Player |
| 3   | LightboxMediaViewer | On close | Stops playing video |

## 3. Metrics

The following events will need to be tracked with IBM's IDA Stats event tracking

| Event name | Event code | Description |
| --- | --- | --- |
| launched | 0 | Video launched from inline or text link state and started playing |
| paused | 1 | Video playback has been paused |
| played | 2 | Video playback has started (should trigger playerState “launched” if the currentTime is 0) |

> | ended | 3 | Video playback has reached the end |

Reviewer: I also kind of wonder about the relevance of this anymore. All of our components should be enabled for metric tracking. I wonder if this should be more along the lines of available events instead of what Google analytics is doing with those events?

Author: They made no sense to me! Would love to remove. Just weren't sure what it's for.
